### ISSUE TITLE
the future of rust-ipfs

### ISSUE THREAD

**Issue by dvc94ch** [ipfs-rust/rust-ipfs#136]
The purpose of this issue is to explain and solicit feedback on a long term vision for the project. @vmx, @aphelionz, @koivunej, @rklaehn 

Current ipfs implementations have a monolithic architecture. To make rust-ipfs attractive as core component of new os's and embedded systems, I propose the following microservice architecture:

ipld or block daemon: this system daemon has a minimal api surface and is responsible for maintaining system integrity and being a good p2p citizen. this involves the following tasks:

- it chown's a system wide directory tentatively called `/ipfs` and is the only process that can write to that directory
- clients can read blocks from the filesystem by reading `/ipfs/block/{cid}`
- to write blocks on windows a http request is made, as the daemon needs to ensure that the cid matches the data. on unix systems unix domain sockets are used
- to decrease the write load on the daemon a more complicated system could be built where the client writes the block to a temporary location, which the daemon monitors and copies into the system wide repo. note that this would require the daemon to be running with chown capabilities and that reads would require two file system lookups
- serving blocks to peers via bitswap
- applications create a directory in `/ipfs/pin/{user-name}/{application-name}` which they own
- to pin a block they create a symlink from `/ipfs/per-user/{user-name}/per-application/{application-name}/pins/{cid}` -> `/ipfs/blocks/{cid}`. a file lock is aquired on `/ipfs/.lock` to prevent races with the garbage collector.
- the garbage collector is run periodically or on demand. it reads all the application directories and creates a list of gc roots. all the blocks are scraped for strong links (introducing strong/weak links) which are put in a list in of used blocks. all blocks not in that list are put in the garbage list. this list is sorted in reverse topological order and the blocks are removed from the store. care needs to be taken to ensure that the garbage collection process is concurrent and system integrity is maintained (example problem scenario application pins block after it's in the remove list). these issues are documented and solved in literature (@eelco dolstra's phd)

ipns-daemon: to be discussed in the future in a separate issue

ipfs-compatibility-daemon: provides the ipfs http api using the ipld and ipns daemons and libraries like libp2p, libipld, ipld-collections and unixfs.

fuse file system: Uses the unixfs library to provide a file system on unix'es

nginx proxy: Serves html from the fuse file system

**Comment by aphelionz** [comment]
@dvc94ch May I have your permission to edit the OP (not destroying any information but just formatting and organizing it) as the conversation continues in the comments?

**Comment by dvc94ch** [comment]
Yes, go ahead

**Comment by vmx** [comment]
That sounds good to me. I'd expect that being one implementation of such a daemon, others might want to use a database/key-value store to store the blocks. But that shouldn't make a difference to the API.

**Comment by dvc94ch** [comment]
using a database/key-value store makes sense if you have a single user system, in which case you can embed something like sled or some other db into the application. the downside is that blocks can't be shared between applications, but for an embedded system running only one application it makes perfect sense. also for running ipfs in the browser it makes sense as due to sandboxing blocks can't be shared.

which api do you mean? the main point I'm trying to make is I dislike the monolithic architecture of existing ipfs implementations. there is no reason for example why an application can't use libp2p to do pubsub instead of using a http api.

**Comment by Gozala** [comment]
Some notes:

> it chown's a system wide directory tentatively called `/ipfs` and is the only process that can write to that directory.

That is probably not a good name as ipfs fuse thing uses `/ipfs` and `/ipns` to materialize local views into network.

> clients can read blocks from the filesystem by reading /ipfs/block/{cid}

What is the client in this context ? You mean arbitrary process can read files from here ? If so I think it might be better to avaid expose internal data layout to the clients for the following reasons:

- I suspect data layout / encoding may change over time as more performance optimizations are considered.
- In some OSs that may not be possible or different thing might make more sense e.g. iOS sandboxing would prevent such a thing.
- OS's seem to move towards better more granular sandboxing that this seem to contradict.

> to write blocks on windows a http request is made, as the daemon needs to ensure that the cid matches the data. on unix systems unix domain sockets are used

Maybe better alternative would be to have notion of child process that internally use domain socket, http socket, maybe operate with FS directly (kinda like sqlite) but expose API over IPC.

> to decrease the write load on the daemon a more complicated system could be built where the client writes the block to a temporary location, which the daemon monitors and copies into the system wide repo. note that this would require the daemon to be running with chown capabilities and that reads would require two file system lookups

**Just spitballing here**: What if each client had it's own "staging area" where it can read / write without extra overhead. Changes that could be committed to the system wide repo by notifying daemon.

Maybe techniques used by sqlite could be employed to even remove the need for separate daemon process for doing commits.

**Comment by dvc94ch** [comment]
> What is the client in this context ? You mean arbitrary process can read files from here ?

The idea is that there is a library for doing this. Potentially it could be a dynamic library that get's installed with the daemon, so changes to the layout do not affect applications.

> Just spitballing here: What if each client had it's own "staging area" where it can read / write without extra overhead. Changes that could be committed to the system wide repo by notifying daemon.

Exactly this is what I was proposing.

> Maybe techniques used by sqlite could be employed to even remove the need for separate daemon process for doing commits.

As mentioned earlier, to maintain system integrity this is not possible. By maintaining system integrity I mean the following:

- prevent an application from breaking the content addressing invariant (hash must match data)
- prevent an application from deleting a block required by another application

**Comment by Gozala** [comment]
> serving blocks to peers via bitswap

Does it needs to be it's job ? It seems like a separate task, another client of the daemon. The reason I say this is because:

1. I can imagine (and in fact have use cases) of wanting to store some blocks that I do not want to share with network.
2. Use cases where I want to be selective about who I share which blocks with. Private group blocks that are shared with only peers of that group.

**Comment by Gozala** [comment]
> As mentioned earlier, to maintain system integrity this is not possible. By maintaining system integrity I mean the following:

We had a race there 😂

> - prevent an application from breaking the content addressing invariant (hash must match data)
> - prevent an application from deleting a block required by another application

Integrity could be enforced through that client process.

**Comment by dvc94ch** [comment]
That could be a separate service, but it would require the daemon to notify the service when a block is added. This is something I was considering also as graphsync/bitswap could be two separate services.

> I can imagine (and in fact have use cases) of wanting to store some blocks that I do not want to share with network.

I think the better solution may be to require private data to be encrypted. This makes selecting what is private or not easier, and it allows for syncing data between your laptop and phone for example. The application should make sure that no sensitive data is leaked.

> Integrity could be enforced through that client process. But that would probably conflict with my sandboxing argument.

The only way to enforce integrity is through the kernel (unix file permissions for example). Just providing a client is not enough, as anyone could write a client that does not follow the "rules"

**Comment by Gozala** [comment]
> I think the better solution may be to require private data to be encrypted. This makes selecting what is private or not easier, and it allows for syncing data between your laptop and phone for example. The application should make sure that no sensitive data is leaked.

I would argue you need both! e.g adversary can identify conspirators without necessary having access to the data itself.

**Comment by dvc94ch** [comment]
Would you like to elaborate on how/who makes the decision which data is private and which not?

Also be interested in more details about risk of publicly exposing encrypted data.

**Comment by Gozala** [comment]
> The only way to enforce integrity is through the kernel (unix file permissions for example). Just providing a client is not enough, as anyone could write a client that does not follow the "rules"

👍 

I was thinking that client would run under different unix user with different permissions. However I have to admit, I'm starting to feel out of my depth here. Nor do I if this portable across OSes. Never the less I did mention it as throwing all the options at a table my lead to some new insights.

Overall I think there is something to the idea of each client operating in their own namespace and submitting changes to the global repo as that provides independence of operation (even if daemon is not running) and allows eventual consistency (using daemon when it is running).

**Comment by Gozala** [comment]
> Would you like to elaborate on how/who makes the decision which data is private and which not?

Would it make sense to discuss this on a different thread ?

> Also be interested in more details about risk of publicly exposing encrypted data.

@gpestana gave a talk about metadata leakage at http://dtn.is/ that I think summarized it fairly well https://www.youtube.com/watch?v=lGk6OErtDFY&t=25s

**Comment by dvc94ch** [comment]
I completely agree with you, we're basically talking about ironing out the details, and would love if more people would contribute to the discussion.

> Would it make sense to discuss this on a different thread ?

Sounds good!

**Comment by dvc94ch** [comment]
Thinking about privacy issues, I think we have them anyway. Does exposing encrypted private data make things worse? Don't know, but in general people are willing to give up privacy to large companies, which sell the data to anyone with deep enough pockets. Making the data available to everyone means that small companies have the same resources available to them as large ones, don't see that being a bad thing, but that's obviously an opinion...

**Comment by alexjg** [comment]
This is a somewhat general comment but I'm personally interested in controlling who I share data with, even if it's encrypted. My main concern is limiting the damage if keys are compromised.

**Comment by dvc94ch** [comment]
Good point. If you or @Gozala want to open an issue to discuss the requirements, that would be great.

Some approaches might be:
- applications store private blocks in a local store
- the cid encodes if the data is public or private

Some questions to answer are:
- is the distinction between public/private enough
- what's the responsibility of the daemon and what's up to the application

**Comment by aphelionz** [comment]
FWIW, with the exception of a private IPFS network via Swarm key, it's historically been the case that encryption / ACL was always delegated to the application layer. Correct me if I'm wrong, @vmx 

My speculation is that this is, in part, probably to avoid the philosophical / moral / ethical underpinnings of the discussion. Either way, a new issue was opened here: https://github.com/ipfs-rust/rust-ipfs/issues/138
